[Watcher] Support for adding actions to threshold watch

alisonelizabeth opened the pull request:
This PR adds support for adding actions via threshold alert. 

- [x] Logging support
- [x] Slack support
- [x] Jira support
- [x] Webhook support
- [x] Email support
- [x] Index support
- [x] PagerDuty support

Other action items:

- [x] Field validation for each action
- [x] Remove hipchat support
- [x] Server-side support for missing actions (Jira, PagerDuty, Index & Webhook)

---
### Instructions
<details>
<summary>Instructions for setting up accounts to test actions.</summary>
<br>

Index, webhook and logging actions can be tested without any configuration in ES.

All other actions need at least one configured account to work.

#### Slack

[Follow documentation here](https://www.elastic.co/guide/en/elastic-stack-overview/current/actions-slack.html#configuring-slack).

I used my own account and created a dummy channel: `#alison-watcher-test`. I plan on archiving after testing is complete.

#### PagerDuty

[Follow documentation here](https://www.elastic.co/guide/en/elastic-stack-overview/current/actions-pagerduty.html#configuring-pagerduty).

#### Jira

[Follow documentation here](https://www.elastic.co/guide/en/elastic-stack-overview/current/actions-jira.html#configuring-jira).

Please DM me for details on a Jira account to use.

#### Email

[Follow documentation here](https://www.elastic.co/guide/en/elastic-stack-overview/current/actions-email.html#configuring-email).

</details>

---
### Screenshots

Actions panel with no actions selected:
<img width="1181" alt="Screen Shot 2019-04-24 at 2 15 00 PM" src="https://user-images.githubusercontent.com/5226211/56683485-a5416080-669b-11e9-84a3-e69f0cb441cf.png">

Actions dropdown:
<img width="404" alt="Screen Shot 2019-04-23 at 2 25 24 PM" src="https://user-images.githubusercontent.com/5226211/56606188-ae66fa80-65d3-11e9-84a3-3d606a5c1851.png">

Email action:
<img width="1136" alt="Screen Shot 2019-04-29 at 11 50 23 AM" src="https://user-images.githubusercontent.com/5226211/56908920-40a84c00-6a75-11e9-885b-d7a45be20a07.png">

Logging action:
<img width="1119" alt="Screen Shot 2019-04-29 at 11 50 32 AM" src="https://user-images.githubusercontent.com/5226211/56908945-49991d80-6a75-11e9-8d25-a69b92ef32fa.png">

Slack action:
<img width="1119" alt="Screen Shot 2019-04-29 at 11 50 40 AM" src="https://user-images.githubusercontent.com/5226211/56908957-5158c200-6a75-11e9-8d1a-b458d7bdf900.png">

Webhook action:
<img width="1152" alt="Screen Shot 2019-04-29 at 11 51 02 AM" src="https://user-images.githubusercontent.com/5226211/56908973-59186680-6a75-11e9-8da3-03a5a0fda198.png">

Index action:
<img width="1131" alt="Screen Shot 2019-04-29 at 11 51 19 AM" src="https://user-images.githubusercontent.com/5226211/56908991-5fa6de00-6a75-11e9-9c15-afae7c92be61.png">

PagerDuty action:
<img width="1113" alt="Screen Shot 2019-04-29 at 11 51 35 AM" src="https://user-images.githubusercontent.com/5226211/56908999-659cbf00-6a75-11e9-8bb5-1e28b85f7690.png">

Jira action:
<img width="1069" alt="Screen Shot 2019-05-01 at 9 08 24 AM" src="https://user-images.githubusercontent.com/5226211/57020688-17fc8f80-6bf8-11e9-9b4a-9279df1a7957.png">

Action with field error:
<img width="1061" alt="Screen Shot 2019-04-29 at 11 54 15 AM" src="https://user-images.githubusercontent.com/5226211/56909095-8bc25f00-6a75-11e9-9e1f-c29b15243c63.png">

Advanced watch with invalid action:
<img width="1136" alt="Screen Shot 2019-04-24 at 2 00 33 PM" src="https://user-images.githubusercontent.com/5226211/56683524-b8ecc700-669b-11e9-98c5-116e97353993.png">

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9860/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10120/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10321/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10415/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10559/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10569/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10601/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10832/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10844/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10856/)

elasticmachine commented:
Pinging @elastic/es-ui

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11052/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11056/)

alisonelizabeth commented:
@cjcenizal Mind taking a first pass at reviewing this? I still have a fair amount of testing to do, but overall functionality is working.

I have a few thoughts/concerns, welcome any feedback here:
- There’s probably some room for improvement with the UI, so please let me know if you have any suggestions.
- The existing validation logic for actions is a little complex, both on the client and server. There was sort of a one-off validation that existed for slack actions that was used for the advanced watch. Not sure that really makes sense anymore. But, also, in general, there seems to be a lot going on :) (For history, I was referencing this PR: https://github.com/elastic/kibana/pull/23887. I need to make sure we’re still catching all of these scenarios, but also wondering if there’s a better approach to take.)
- There’s a good amount of boilerplate code needed to stick with the existing pattern.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11062/)

cjcenizal commented on x-pack/plugins/watcher/public/sections/watch_edit/components/threshold_watch_action_dropdown.tsx:
```diff
@@ -0,0 +1,98 @@
+  }, []);
+  const actionOptions =
+    actions &&
+    actions.map((action: ActionType) => {
```
You can simplify this slightly by ensuring `actionOptions` is defined as an array:

```tsx
  const actionOptions = actions ?
    actions.map((action: ActionType) => {
      const description = action.isEnabled ? action.selectMessage : disabledMessage;
      return {
        value: action.type,
        inputDisplay: action.typeName,
        disabled: !action.isEnabled,
        dropdownDisplay: (
          <EuiFlexGroup>
            <EuiFlexItem grow={false} style={{ alignSelf: 'center' }}>
              <EuiIcon type={action.iconClass} />
            </EuiFlexItem>
            <EuiFlexItem grow={false}>
              <strong>{action.typeName}</strong>
              <EuiSpacer size="xs" />
              <EuiText size="s" color="subdued">
                <p className="euiTextColor--subdued">{description}</p>
              </EuiText>
            </EuiFlexItem>
          </EuiFlexGroup>
        ),
      };
    }) : [];
    
  const actionOptionsWithEmptyValue = [firstActionOption, ...actionOptions];
```

alisonelizabeth replied:
fixed

cjcenizal commented on x-pack/plugins/watcher/public/sections/watch_edit/watch_edit_actions.ts:
```diff
@@ -75,21 +75,30 @@ export async function saveWatch(watch: BaseWatch, licenseService: any) {
 }
 
 export async function validateActionsAndSaveWatch(watch: BaseWatch, licenseService: any) {
-  const { warning } = watch.validate();
```
Do you know what this used to do? I'm just curious about how the UX has changed with the removal of this logic.

alisonelizabeth replied:
The original code displayed this modal if the `to` field was missing from a slack action.

<img width="811" alt="Screen Shot 2019-04-11 at 8 48 33 AM" src="https://user-images.githubusercontent.com/5226211/56746954-2441a200-674b-11e9-8da8-e73d3de607ec.png">

However, when setting up an integration with Slack, it seems it's required to provide a "to" value.

<img width="1056" alt="Screen Shot 2019-04-25 at 11 17 25 AM" src="https://user-images.githubusercontent.com/5226211/56747317-bba6f500-674b-11e9-8118-2edf335a157f.png">

If an option is not selected, the save button is disabled.

<img width="331" alt="Screen Shot 2019-04-25 at 11 15 46 AM" src="https://user-images.githubusercontent.com/5226211/56747316-bb0e5e80-674b-11e9-876c-405f4e5bb47f.png">

The value can be overridden in ES, but if it's not defined, the notification will go to the channel specified in Slack.

Also see: https://github.com/elastic/kibana/issues/18531#issuecomment-384127356

@sebelga I know I talked to you about this originally. Does the above make sense? Or I am missing another case?

cjcenizal replied:
Cool! If I understand correctly, we originally thought that a Slack watch saved without a `to` field either on the watch itself or in the ES settings would be essentially useless, because it wouldn't end up doing anything when triggered. However, we've learned that Slack allows you to use a Slack watch even if there is no `to` field specified on our end, rendering this warning moot/incorrect.

sebelga replied:
I am not sure this is correct. From the comment

> The way I setup my Slack Webhook for this watch, it has a default "to" on the slack side.

So a "to" _is_ required. The thing is, the user can set it multiple ways. The warning that was put in place is: "we can't validate that you have set properly a "to" field, be aware that the watch might not work if you haven't specified it outside our UI".

This validation was useful if the user enters "Create advanced watch", add a JSON payload of the watch with a Slack action, but without a "to" field.

alisonelizabeth replied:
@cjcenizal that's correct.

@sebelga As far as I can tell, you would have to **at minimum** set a "to" via Slack. So while you could override it via ES, I think you would **always** have to set it in Slack (see screenshots above) in order to generate the webhook url. Does that make sense?

Basing this on the docs here and my own setup experience: https://www.elastic.co/guide/en/elastic-stack-overview/7.0/actions-slack.html#configuring-slack

Also, note in the docs:

> You can no longer configure Slack accounts using elasticsearch.yml settings. Please use Elasticsearch’s secure keystore method instead.

alisonelizabeth commented:
@cjcenizal thanks for the initial review!

1. So it seemed like the actions weren't fully implemented. Some additional action models were added to the client, but not on the server via this PR: https://github.com/elastic/kibana/pull/30043. I also noticed `index.action.js`, for example, did not have the correct fields (https://github.com/elastic/kibana/pull/35175/files#diff-57ff8b6a2263badfe04d61e472292927). It looked like maybe it had accidentally been copied from the webhook action. I'll continue to do some testing, but everything seemed to be working as expected with my changes. @sebelga do you think you could also help verify this area as well?
2. The `action_defaults` files were responsible for adding some default values to the action fields, e.g., for logging, the default logging text is: ```"Watch [{{ctx.metadata.name}}] has exceeded the threshold."``` I thought it was all hard-coded text, so I moved it to each action’s model. However, now that I’m looking at the files again, I need to go back and verify the email action. It looks like it is also setting the “to” email address if `xPack:defaultAdminEmail` is set.
3. I deleted files related to the settings service in favor of moving the settings request to the `api.ts` file. I think all of the other deleted files are related to the comments in 1,2
4. I definitely agree on the folder structure being hard to follow. What do you think about this? *Note:* I also moved `watch_context` out of the components dir. 

<img width="624" alt="Screen Shot 2019-04-25 at 8 12 28 AM" src="https://user-images.githubusercontent.com/5226211/56737405-5dbce200-6738-11e9-8113-2e4ce06c6f2c.png">

5. Good call on the spacing. I think I had that in there, but accidentally took it out again when I was cleaning up some things.

alisonelizabeth commented on x-pack/plugins/watcher/public/sections/watch_edit/components/logging_action_fields.tsx:
```diff
@@ -0,0 +1,57 @@
+          onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
+            editAction({ key: 'text', value: e.target.value });
+          }}
+          onBlur={() => {
```
This was added as part of the validation implementation. If a field is required, it won’t be marked as so until after the user leaves an input field.

alisonelizabeth replied:
side note - i noticed for the SR work, the fields are validated after the user submits, rather than immediately. is there a preferred approach?

cjcenizal replied:
> side note - i noticed for the SR work, the fields are validated after the user submits, rather than immediately. is there a preferred approach?

Form validation UX is tough to get right and I think what we're doing elsewhere works pretty well but I'm not at all confident that it's ideal.I don't think it matters much that it works slightly differently than we have in other UIs. Given that the UX seems to work pretty well, I don't think we need to spend more time refining it in this iteration.

>  If a field is required, it won’t be marked as so until after the user leaves an input field.

Could you add a comment describing this intention here and in other places where we use an onBlur this way?

alisonelizabeth commented on x-pack/plugins/watcher/public/sections/watch_edit/components/email_action_fields.tsx:
```diff
@@ -0,0 +1,120 @@
+  hasErrors,
+}) => {
+  const { to, subject, body } = action;
+  const [toOptions, setToOptions] = useState<Array<{ label: string }>>([]);
```
good point, much cleaner!

alisonelizabeth replied:
fixed

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11236/)

alisonelizabeth commented on x-pack/plugins/watcher/public/sections/watch_edit/components/threshold_watch_action_accordion.tsx:
```diff
@@ -0,0 +1,172 @@
+              }}
+            />
+            <EuiButton
+              fill
```
fixed

alisonelizabeth commented on x-pack/plugins/watcher/public/sections/watch_edit/components/email_action_fields.tsx:
```diff
@@ -0,0 +1,120 @@
+          }}
+        />
+      </ErrableFormRow>
+      <EuiFormRow
```
updated

alisonelizabeth commented on x-pack/plugins/watcher/public/sections/watch_edit/components/json_watch_edit_form.tsx:
```diff
@@ -27,28 +27,47 @@ import { LicenseServiceContext } from '../../../license_service_context';
   const { errors } = watch.validate();
   const hasErrors = !!Object.keys(errors).find(errorKey => errors[errorKey].length >= 1);
-  const licenseService = useContext(LicenseServiceContext);
+  const [validationResult, setValidationResult] = useState<{
```
updated

alisonelizabeth commented on x-pack/plugins/watcher/public/sections/watch_edit/components/slack_action_fields.tsx:
```diff
@@ -0,0 +1,101 @@
+  hasErrors,
+}) => {
+  const { text, to } = action;
+  const [toOptions, setToOptions] = useState<Array<{ label: string }>>([]);
```
fixed

alisonelizabeth commented on x-pack/plugins/watcher/public/sections/watch_edit/components/threshold_watch_action_accordion.tsx:
```diff
@@ -0,0 +1,172 @@
+import { JiraActionFields } from './jira_action_fields';
+import { executeWatch } from '../../../lib/api';
+
+const ActionFieldsComponent = {
```
renamed

alisonelizabeth commented:
@cjcenizal I pushed some changes based on your review and a couple other bugs I noticed while testing.

Re:
> The action_defaults files were responsible for adding some default values to the action fields, e.g., for logging, the default logging text is: "Watch [{{ctx.metadata.name}}] has exceeded the threshold." I thought it was all hard-coded text, so I moved it to each action’s model. However, now that I’m looking at the files again, I need to go back and verify the email action. It looks like it is also setting the “to” email address if xPack:defaultAdminEmail is set.

I fixed this by calling `chrome.getUiSettingsClient()` in the email-action model to check if there is a default email address set.

> I definitely agree on the folder structure being hard to follow. What do you think about this? Note: I also moved watch_context out of the components dir.

I went ahead and made changes to the folder structure as shown in my above comment. Let me know what you think.

-----

FYI: I need to do more thorough testing for Jira and PagerDuty actions, but I think logging, index, email and Slack are in good shape now.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11259/)

alisonelizabeth commented:
@bmcconaghy / @cjcenizal Something else I came across while testing…

The current code was calling the cluster settings API to determine if an action was enabled or not for this dropdown menu. 

<img width="426" alt="Screen Shot 2019-04-23 at 2 31 43 PM" src="https://user-images.githubusercontent.com/5226211/56606641-ae1b2f00-65d4-11e9-9a30-999190b615a3.png">

However, I don’t think this really achieves what we want, especially now that we have added support for Jira and PagerDuty. Technically, for Jira and PagerDuty (and Slack), a user only has to set some values via the keystore to configure. In this scenario, there would not be any values under the `xpack.notificaton` property in `elasticsearch.yml` that we’re checking against. Therefore, it would be marked as “not enabled” even though it is. 

It looks like this may have come up before with Slack, and as a workaround, Slack is “enabled” by default in the code. I supposed we could do the same for Jira and PagerDuty, but in that case, the only thing this code is really checking for is email. Thoughts?

**Code in question:** https://github.com/alisonelizabeth/kibana/blob/threshold-watch-actions/x-pack/plugins/watcher/server/models/settings/settings.js

yaronp68 commented:
@alisonelizabeth Great work. I find the action combo too much to the right (compared to other fields)
![image](https://user-images.githubusercontent.com/1367657/56823676-70c0d680-684c-11e9-8222-1fb78eeff190.png)

alisonelizabeth commented:
Thanks, @yaronp68! It's currently right-aligned. I tried to replicate what existing previously, but I'm not sold on it either. Do you have any suggestions?

I wonder if it would help if there was more separation between the different sections? Another idea might be to just show the actions dropdown initially, and only render the text if an action has been added rather than having `Will perform 0 actions...` as a default. Then, maybe the text could sit underneath the dropdown, or something to that effect. Thoughts?

alisonelizabeth commented:
Had a conversation with @bmcconaghy about the actions dropdown. Updated the code to work as follows:

- We will continue to check the yml file to attempt to determine if an action is enabled or not.
- For actions that can technically be set up through the keystore only (Slack, Jira and PagerDuty), we will not disable it in the dropdown and instead render a warning callout indicating that the account may not be set up, with a link to the configuration docs.

<img width="417" alt="Screen Shot 2019-04-26 at 4 10 06 PM" src="https://user-images.githubusercontent.com/5226211/56834676-13bf2380-6840-11e9-8596-b69f72c2e11e.png">

[
<img width="1153" alt="Screen Shot 2019-04-26 at 4 10 14 PM" src="https://user-images.githubusercontent.com/5226211/56834689-1de12200-6840-11e9-8533-6393541f876f.png">
](url)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11423/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11432/)

sebelga commented:
Great job @alisonelizabeth ! For now, I only did a local test of the UI (haven't reviewed the code yet), but here are some comments on the UI. 

- I get a EUI ComboBoxOptionsList warning in the Chrome console when type the index name, has this bug been raised to them?

<img width="793" alt="Screen Shot 2019-04-29 at 16 49 08" src="https://user-images.githubusercontent.com/2854616/56905025-c93be200-6a9f-11e9-8d5d-4cded94eb4fb.png">

- In other apps we've decided to only show validation _after_ the user has submitted the form so we don't "bother" him too early while he sets the form correctly. Would it be something easy to add here?

- I wonder if we shouldn't change the "Add an action" to a primary button with a "+" icon in it. When clicking on it, the dropdown that is currently in place would replace the button, and there would be "cancel" link button right next to it to revert. What do you think? (cc @cjcenizal )

- For the Webhook action, do we need to separate the URL in 3 fields? Couldn't we let the user simply enter

```
http://www.mydomain.com:7000/some-path
// or
https://mydomain.com:7000/some-path
```

and extract the fields from there? If it's complex we could leave it for a second pass but, as a user, I would expect to be able to copy paste a URL in one field.

One last thing, could you add in the PR description how we can test the integration (email, Slack, Jira, Pager duty)? I'm sure I would find it in the docs somewhere but it would be much easier to copy paste your `elasticsearch.yml` file 😊

alisonelizabeth commented:
Thanks, @sebelga!

I updated the PR with links to some setup instructions. You can reach out to me regarding Jira, and of course if you have any other questions!

- I didn’t work on this particular part, but noticed that warning as well. I’ve made a note of it and will look into it as a follow-up.
- I noticed that too. Left a comment here: https://github.com/elastic/kibana/pull/35175#discussion_r279026774. This is how the validation is working in all of the new watcher code currently. Perhaps we can address in a separate PR to align with other apps. 
- I’m open to changing the design here. Do you know if there’s an example somewhere that I could take a look at where we are following the pattern you described?
- That’s a good idea. I’ll look into it.

cjcenizal commented:
> I wonder if we shouldn't change the "Add an action" to a primary button with a "+" icon in it. When clicking on it, the dropdown that is currently in place would replace the button, and there would be "cancel" link button right next to it to revert. 

I think the idea of using a button is a good one. I'd suggest using the same pattern Jen used for [rollup index pattern creation](https://github.com/elastic/kibana/pull/24805).

cjcenizal commented:
Retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/11711/)

sebelga commented on x-pack/plugins/watcher/public/sections/watch_edit/components/threshold_watch_edit/threshold_watch_action_dropdown.tsx:
```diff
@@ -0,0 +1,100 @@
+  const allActionTypes = Action.getActionTypes();
+
+  const actions = Object.keys(allActionTypes).map(actionKey => {
+    const { typeName, iconClass, selectMessage } = allActionTypes[actionKey];
```
nit: you could simplify with `Object.entries`

```js
const allActionTypes = Action.getActionTypes() as Record<string, any>;

const actions = Object.entries(allActionTypes).map(([type, { typeName, iconClass, selectMessage }]) => ({
  type,
  typeName,
  iconClass,
  selectMessage,
  isEnabled: settings ? settings.actionTypes[type].enabled : true,
}));
```

alisonelizabeth replied:
thanks - updated

alisonelizabeth commented on x-pack/plugins/watcher/public/models/action/email_action.js:
```diff
@@ -7,15 +7,48 @@
+    if (!this.to || !this.to.length) {
+      errors.to.push(
+        i18n.translate('xpack.watcher.watchActions.email.emailRecipientIsRequiredValidationMessage', {
+          defaultMessage: 'To email address is required.',
```
Good point. Updated, but also plan on doing a copy review with gail.

alisonelizabeth commented on x-pack/plugins/watcher/public/models/action/slack_action.js:
```diff
@@ -4,99 +4,71 @@
   validate() {
-    const errors = [];
-
-    if (!this.to.length) {
```
see comment regarding this above. if you still have concerns, please let me know!

alisonelizabeth commented on x-pack/plugins/watcher/public/models/watch/base_watch.js:
```diff
@@ -135,44 +131,6 @@ export class BaseWatch {
-   * Client validation of the Watch.
-   * Currently we are *only* validating the Watch "Actions"
-   */
-  validate() {
```
There is still validation, but it was added as part of the specific watch (see: `models/watch/threshold_watch.js`, `models/watch/json_watch.js`). Each action model also has it's own validation.

alisonelizabeth commented on x-pack/plugins/watcher/public/sections/watch_edit/components/threshold_watch_edit/action_fields/webhook_action_fields.tsx:
```diff
@@ -0,0 +1,178 @@
+          fullWidth
+          name="method"
+          value={method || 'get'}
+          options={[
```
yep, much cleaner. fixed.

alisonelizabeth commented on x-pack/plugins/watcher/public/sections/watch_edit/components/threshold_watch_edit/threshold_watch_action_accordion.tsx:
```diff
@@ -0,0 +1,229 @@
+import { executeWatch } from '../../../../lib/api';
+import { watchActionsConfigurationMap } from '../../../../lib/documentation_links';
+
+const ActionFieldsComponentMap = {
```
Yeah, my mistake here. Good catch - fixed!

alisonelizabeth commented on x-pack/plugins/watcher/public/sections/watch_edit/components/threshold_watch_edit/threshold_watch_action_accordion.tsx:
```diff
@@ -0,0 +1,229 @@
+    name: string;
+    iconClass: 'loggingApp' | 'logoWebhook' | 'logoSlack' | 'apps' | 'email' | 'indexOpen';
+  }) => (
+    <Fragment>
```
haha. thanks for the catch - fixed. i originally had this set up a little differently where it was necessary and forgot to remove.

alisonelizabeth commented on x-pack/plugins/watcher/public/sections/watch_edit/components/threshold_watch_edit/threshold_watch_action_accordion.tsx:
```diff
@@ -0,0 +1,229 @@
+          buttonContentClassName="euiAccordionForm__button"
+          buttonContent={<ButtonContent name={action.typeName} iconClass={action.iconClass} />}
+          extraAction={
+            <DeleteIcon
```
good point, fixed.

alisonelizabeth commented on x-pack/plugins/watcher/public/sections/watch_edit/components/threshold_watch_edit/threshold_watch_action_accordion.tsx:
```diff
@@ -0,0 +1,229 @@
+  const { watch, setWatchProperty } = useContext(WatchContext);
+  const { actions } = watch;
+
+  const ButtonContent = ({
```
good point. i decided to directly declare this for the `buttonContent` prop

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12006/)

alisonelizabeth commented:
@sebelga, thanks for your review! I believe I addressed all of your comments. Can you take another look when you get a chance?

I’d like to address the design issues in a follow-up PR if that works for you (changing the actions dropdown to a button, and updating the webhook form to have one field for host/port/path)

I opened an issue against EUI regarding the ComboBox, but realized that it is working as intended. We might need to rethink how we’re using it. See: https://github.com/elastic/eui/issues/1901.

I tested the scenario mentioned in #23887. It works as expected in edit mode. However, it looks like we are missing the error column and corresponding modal on the details pages (see screenshots of old UI below). This is unrelated to my changes, so I think this should be addressed in a separate PR. FYI @cjcenizal 

<img width="1014" alt="Screen Shot 2019-04-29 at 4 40 20 PM" src="https://user-images.githubusercontent.com/5226211/57022306-c0145780-6bfc-11e9-8938-7b8bfaf186c7.png">

<img width="685" alt="Screen Shot 2019-04-29 at 4 40 12 PM" src="https://user-images.githubusercontent.com/5226211/57022314-c73b6580-6bfc-11e9-86cd-bde96a5add3a.png">

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12009/)

alisonelizabeth commented:
retest

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12053/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12521/)
